client-testkit/shared/src/main/scala/org/http4s/client/testkit/WSTestClient.scala

```diff
@@ -0,0 +1,144 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s
+package client
+package testkit
+
+import cats.Applicative
+import cats.Foldable
+import cats.MonadThrow
+import cats.effect.Concurrent
+import cats.effect.Resource
+import cats.effect.std.Queue
+import cats.implicits._
+import fs2.Stream
+import org.http4s.HttpApp
+import org.http4s.client.websocket.WSClient
+import org.http4s.client.websocket.WSConnection
+import org.http4s.client.websocket.WSFrame
+import org.http4s.server.websocket.WebSocketBuilder2
+import org.http4s.websocket.WebSocketCombinedPipe
+import org.http4s.websocket.WebSocketFrame
+import org.http4s.websocket.WebSocketSeparatePipe
+
+object WSTestClient {
+
+  def fromHttpWebSocketApp[F[_]: Concurrent](
+      f: WebSocketBuilder2[F] => HttpApp[F]
+  ): F[WSClient[F]] =
+    fromHttpWebSocketApp(respondToPings = true)(f)
+
+  /** Creates a WSClient from the specified [[HttpApp]].
+    * [[org.http4s.server.websocket.WebSocketBuilder2]] is used for specifying the WebSocket connection.
+    * Useful for generating pre-determined responses for requests in testing.
+    *
+    * @param respondToPings if true, the client will respond to ping frames with a pong frame.
+    * @param f a function that takes a [[org.http4s.server.websocket.WebSocketBuilder2]] and returns an [[HttpApp]].
+    */
+  def fromHttpWebSocketApp[F[_]: Concurrent](
+      respondToPings: Boolean
+  )(f: WebSocketBuilder2[F] => HttpApp[F]): F[WSClient[F]] =
+    for {
+      wsb <- WebSocketBuilder2[F]
+      app = f(wsb)
+    } yield WSClient[F](respondToPings) { req =>
+      Resource
+        .eval(app.run(Request[F](method = req.method, uri = req.uri, headers = req.headers)))
+        .flatMap { response =>
+          response.attributes.lookup(wsb.webSocketKey) match {
+            case Some(context) =>
+              context.webSocket match {
+                case WebSocketSeparatePipe(s, r, c) =>
+                  Resource
+                    .pure(new WSConnection[F] {
+
+                      def send(wsf: WSFrame): F[Unit] =
+                        Stream.eval(wsf.toWebSocketFrame).through(r).compile.drain
+
+                      def sendMany[G[_]: Foldable, A <: WSFrame](wsfs: G[A]): F[Unit] =
+                        wsfs.traverse_(send)
+
+                      def receive: F[Option[WSFrame]] =
+                        s.map(_.toWSFrame).head.compile.last
+
+                      def subprotocol: Option[String] = context.subprotocol
+                    })
+                    .onFinalize(c)
+                case wscp: WebSocketCombinedPipe[F] =>
+                  Resource
+                    .liftK(Queue.unbounded[F, WebSocketFrame])
```

Comment:
Leaving this comment just to highlight we are planning to use `unbounded` Queue. Not arguing it's wrong, just fwiw.

Reply:
Oh, interesting point. Probably some sort of `bounded` queue might be more representative of TCP backpressure?

```suggestion
                    .eval(Queue.unbounded[F, WebSocketFrame])
```